patches/tModLoader/Terraria/ModLoader/BossBarLoader.cs

```diff
@@ -0,0 +1,323 @@
+		/// <param name="value">When this method returns, contains the ModBossBar associated with the specified NPC type</param>
+		/// <returns><see langword="true"/> if ModBossBar exists; otherwise, <see langword="false"/>.</returns>
+		public static bool NpcToBossBar(int npcType, out ModBossBar value) {
+			value = null;
+			if (npcToBossBarType.TryGetValue(npcType, out int index))
+				value = bossBars[index];
+			return value != null;
+		}
+
+		/// <summary>
+		/// Matches NPC types to their ModBossBar
+		/// </summary>
+		internal static void BuildReverseLookup() {
+			for (int i = 0; i < bossBars.Count; i++) {
+				ModBossBar bossBar = bossBars[i];
+				var validNPCs = bossBar.ValidNPCs;
+
+				//The same NPC can exist in multiple bars. In that case, only take the first found occurence
+				foreach (int npcType in validNPCs) {
+					if (!npcToBossBarType.ContainsKey(npcType))
+						npcToBossBarType.Add(npcType, i);
+				}
+			}
+		}
+
+		/// <summary>
+		/// Inserts the boss bar style select option into the main and ingame menu under the "Interface" category
+		/// </summary>
+		internal static string InsertMenu(out Action onClick) {
+			string styleText = null;
+			ModBossBarStyle pendingBossBarStyle = null;
+			foreach (ModBossBarStyle bossBarStyle in bossBarStyles) {
+				if (bossBarStyle == CurrentStyle) {
+					styleText = bossBarStyle.DisplayName;
+					break;
+				}
+
+				pendingBossBarStyle = bossBarStyle;
+			}
+
+			if (pendingBossBarStyle == null)
+				pendingBossBarStyle = bossBarStyles.Last();
+
+			if (styleText == null || bossBarStyles.Count == 1)
+				styleText = Language.GetTextValue("tModLoader.BossBarStyleNoOptions");
+
+			onClick = () => SwitchBossBarStyle(pendingBossBarStyle);
+
+			return Language.GetTextValue("tModLoader.BossBarStyle", styleText);
+		}
+
+		public static bool PreDraw(SpriteBatch spriteBatch, BigProgressBarInfo info, ref BossBarDrawParams drawParams) {
+			int index = info.npcIndexToAimAt;
+			if (index < 0 || index > Main.maxNPCs)
+				return false; //Invalid data, abort
+
+			NPC npc = Main.npc[index];
+
+			bool modify = true;
+
+			if (NpcToBossBar(npc.type, out ModBossBar bossBar)) {
+				drawParams.barTexture = GetTexture(bossBar).Value;
+				modify = bossBar.PreDraw(spriteBatch, npc, ref drawParams);
+			}
+
+			foreach(GlobalBossBar globalBossBar in globalBossBars) {
+				modify &= globalBossBar.PreDraw(spriteBatch, npc, ref drawParams);
+			}
+
+			return modify;
+		}
+
+		public static void PostDraw(bool skipped, SpriteBatch spriteBatch, BigProgressBarInfo info, BossBarDrawParams drawParams) {
+			int index = info.npcIndexToAimAt;
+			if (index < 0 || index > Main.maxNPCs)
+				return; //Invalid data, abort
+
+			NPC npc = Main.npc[index];
+
+			if (NpcToBossBar(npc.type, out ModBossBar value))
+				value.PostDraw(skipped, spriteBatch, npc, drawParams);
+
+			foreach (GlobalBossBar globalBossBar in globalBossBars) {
+				globalBossBar.PostDraw(skipped, spriteBatch, npc, drawParams);
+			}
+		}
```

Comment:
```suggestion
		public static void PostDraw(bool skipped, SpriteBatch spriteBatch, BigProgressBarInfo info, BossBarDrawParams drawParams) {
			int index = info.npcIndexToAimAt;
			if (index < 0 || index > Main.maxNPCs)
				return; //Invalid data, abort

			NPC npc = Main.npc[index];

			foreach (GlobalBossBar globalBossBar in globalBossBars) {
				globalBossBar.PostDraw(skipped, spriteBatch, npc, drawParams);
			}

			if (NpcToBossBar(npc.type, out ModBossBar value))
				value.PostDraw(skipped, spriteBatch, npc, drawParams);
		}
```

Reply:
why should Global be first in PostDraw? It's not like this for NPC or Item PostDraw

Reply:
Mirsario told to ping you @Chicken-Bones. He proposes Global `PostDraw` to be before Mod `PostDraw`, unlike other draw methods in TML.

Reply:
I looked at NPCLoader once more and am retracting this. Haven't seen that there's already hooks in which globals are called afterwards on purpose, though I think you told me that.